### ISSUE TITLE
Implement lazy sstable parts loading and caching

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#13219]
HEAD: 25cf3256743d98f1dd34ef0656ee7cc974f87249

*Motivation*
When Scylla starts it loads specific parts of all SStables into the RAM: `Summary`, `Compression Info` (something else, @raphaelsc ?).

And these parts sizes are proportional to the size of the corresponding SStable `Data` component.

These two facts immediately lead to the infamous limitation of the maximum amount of storage attached to a single Scylla node - it must be not greater than x100 of the RAM size on the same node.

This obviously creates a limitation for "large storage" use cases that otherwise would fit Scylla nicely.

*Proposal*
The proposed solution is as old as programming itself:
* Load the content of SStable components above in a lazy way - only when we need to read the corresponding SStable.
* Cache the read content and restrict the maximum total size of memory consumed by the corresponding content in such a cache.
* Evict cached entries using the LRU policy and/or after a specific time when not read.

To those familiar with `scylla` internals the above should immediately resemble our internal prepared statements and authorization related caches.

And indeed we can use the code from https://github.com/scylladb/scylladb/blob/master/cql3/prepared_statements_cache.hh as a reference in order to implement caching of corresponding SStable components.

**Comment by vladzcloudius** [comment]
@denesb FYI

**Comment by denesb** [comment]
Let's look at each component in turn and what it would take for us to evict it from memory. In general the most important thing we need to be able to access the component on disk is the ability to issue random access reads into it, to fetch just the desired entries. This is simple for some and not so simple for others.

### Compression Info

The compression info is just a header and a list of offset pairs (logical offset -> physical offset). We need to read the header in advance to know the offset (in the file) of the offset list. After that we can issue random reads into the file to get just the offset pairs we need. I think in general we could get away with just one additional IO/read if the compression-info is not in memory. Not great, not terrible. We can employ caching like @vladzcloudius suggested, but we may not even need it.

### Summary

While the summary does scale with the size of the sstable, the scaling is not as strict as we can decide to change the sampling ratio and thus reduce the size of the summary. Indeed  I believe we do this already, we tweak the sampling ration as we write the sstable if we discover it is too big. However I think there is a limit to how sparse the summary can be made, as a sparse summary will lead to large index pages, in turn leading to more IO reading the index. However if we are considering not keeping the summary in memory, maybe a very sparse in-memory summary is a good alternative.


Not keeping the summary in memory is a tougher nut to crack because AFAIK the file format doesn't lend itself to random access reads, as the size of entries is not uniform (entries contain partition keys).
There are two options to allow random reads into the file (to fetch a desired entry):
1) Remove the partition key from the summary entry -- just store token + position. This makes the entries of uniform size. I don't think the key is terribly useful anyway. Token collisions are rare so I think we can allow ourselves to not optimize for them.
2) Keep a summary of the summary in memory. I don't even know how this would work. I think option 1 above is just so much better.

Another consideration is that the usual access pattern for the summary is a binary search. If we evict it from memory, we need to make sure we are doing this efficiently. We already have a working solution for on-disk binary search, employed in promoted index, complete with a page cache. I think we could use the same to binary search the summary, but again we need random-access capability for this to work.

### CRC32

This is another component that scales with the size of the sstable. This is only written for non-compressed sstables. We don't use this component at all. Maybe we should but that is for another discussion.

### Fixed size components

We also have a bunch of (more-or-less) fixed-size components:
* Statistics
* Scylla-Metadata
* Filter
* TOC

I don't know how much space these can take up. At a certain point we have to consider not keeping these in memory either. We do already evict part of these (shard ranges). Maybe our memory:disk ratio would already be satisfactory after the large components are evicted from memory.

**Comment by xemul** [comment]
Is Filter really fixed size? AFAIU it should provide some hit/mis ratio and in this case it's size should depend on the size of the data. Or more precisely -- on the number of keys in it

**Comment by denesb** [comment]
> Is Filter really fixed size? AFAIU it should provide some hit/mis ratio and in this case it's size should depend on the size of the data. Or more precisely -- on the number of keys in it

Yes, filter usually scales with the number of keys in the sstable and yes it can get largish. But it is a component we can choose the size of, at the cost of filtering accuracy.
On the other hand, with a cold-storage sstable, you want as accurate filtering as possible, to avoid the expensive false-positive reads. So maybe we should make the filter large and keep it on disk. This would cause some churn in the filtering code as it is currently all synchronous.

**Comment by xemul** [comment]
@avikivity , on one of object-storage meetings you told there was some doc regarding mem:disk ratio. Would you please share one here?

**Comment by vladzcloudius** [comment]
@denesb Don't you think that your ideas https://github.com/scylladb/scylladb/issues/13219#issuecomment-1473180078 deserve a separate RFC GH issue?

Let's summarize what ideas been mentioned so far:

### Idea 1:  (Original idea of this RFC)

Don't change the logic of the current read-path code that assumes that SSTable components are "fast to get". Only change the way their content is stored in the memory - async loading cache instead of an in-memory cache. 

Required changes in the code (according to what was mentioned earlier):
* Implement reading of the corresponding SSTables' content via `loading_cache`.
* Change the corresponding places in the read-path that read that content that are synchronous today to be asynchronous.

Risks:
* Very low since the change is going to utilize a heavily used abstraction that is supposed to be relatively free of bugs and the new code on top of it is pretty trivial.

Possible performance implications: 
* Close to zero and even some improvement is possible. Depending on how we decide to configure those caches only cases when Scylla was OOMing today may trigger "cache pollution" behavior and hence worse-than-optimal performance. (And this is *exactly* how we want it). 

### Idea 2 (https://github.com/scylladb/scylladb/issues/13219#issuecomment-1473180078 - there are actually multiple ideas depending on the specific SSTable component):

Consider not to keep some SSTable components in the memory at all and read them every time we need instead or don't read them at all (CRC32).

General description of required changes in the code (according to what was mentioned earlier):
* Change the corresponding places in the read-path that read that content that are synchronous today to be asynchronous.
* Do some low level optimizations that would allow us real a lot less data from disk.

Risks:
* Considerable: those low level format/configuration changes represent a significant potential risk due to possible implementation bugs or unexpected phenomenas that haven't been triggered so far. 

Possible performance implications:
* Unknown, could potentially be very low or very significant - testing is required.

------------------------------

I think all ideas above are worth exploring. However Idea 1 and Idea(s) 2 are mostly orthogonal and can be developed in parallel or sequentially (see below). Hence I suggest to keep them in separate GH issue because both deserve separate discussions.

Idea 1 seems to be quite conservative and rather simple to implement. And IMO we can get to implementing it without too much discussion.

Idea(s) 2 seem to be more groundbreaking and can require both more discussion and more testing.

Yes, that's right that if we decide not to keep those sstables' components in memory at all (as Ideas 2 propose) then Idea 1 becomes unnecessary. However this may be quite risky and IMO we should consider moving forward in small steps (this is usually the fastest way):

1) Implement Idea 1.
2) Implement Idea(s) 2. If you have (1) it's very easy to simulate (2) - you simply set the corresponding cache size to 0.

**Comment by xemul** [comment]
https://docs.google.com/document/d/1XHrSkTvd0FzjJZcQpVdcKDPE9RowmdjlcVn9MfmQMRI/edit#heading=h.jcr4nczgau14

**Comment by vladzcloudius** [comment]
> We should think about this in 2 ways: a. General improvement of disk/ram ratio Not just sstable/file components. Any improvement is good b. Object store style - need to have > 10x - 50x better ratio with ability to perform worse, assume the data is barely read, so no need to cache it.

@dorlaor This proposal achieves all these goals in a very low price:
1) It reduce the possible OOMs.
2) It removes the Disk:Ram ratio requirement completely (not just "improves it").

**Comment by denesb** [comment]
> I think all ideas above are worth exploring. However Idea 1 and Idea(s) 2 are mostly orthogonal and can be developed in parallel or sequentially (see below). Hence I suggest to keep them in separate GH issues because both deserve separate discussions.
> 
> Idea 1 seems to be quite conservative and rather simple to implement. And IMO we can get to implementing it without too much discussion.

The two ideas you mention above are not orthogonal. Idea1 is Idea2 for CompressionInfo. Note that whether we use caching or not for CompressionInfo is just a small detail *after* we have figured out and implemented reading it from disk.

**Comment by denesb** [comment]
> @denesb Don't you think that your ideas [#13219 (comment)](https://github.com/scylladb/scylladb/issues/13219#issuecomment-1473180078) deserve a separate RFC GH issue?

I think in general it is best to open issues for problems, then discuss ideas to solve it in the comments. In the end, we will choose exactly one solution, implement it and that will fix the issue. If we open separate issues for each idea, we will have to close as "won't fix" all but 1 of them.

**Comment by mykaul** [comment]
> > @denesb Don't you think that your ideas [#13219 (comment)](https://github.com/scylladb/scylladb/issues/13219#issuecomment-1473180078) deserve a separate RFC GH issue?
> 
> I think in general it is best to open issues for problems, then discuss ideas to solve it in the comments. In the end, we will choose exactly one solution, implement it and that will fix the issue. If we open separate issues for each idea, we will have to close as "won't fix" all but 1 of them.

Generally, Github issues are not the best venue for design discussions. We are somewhat lacking a better one though (an open to the world Google Doc? Eh...)

**Comment by dani-tweig** [comment]
@mykaul , why was this issue bumped to P1? Is it meant to be in Q2 content?
@tzach, Is this one of the improved use cases we are aiming in enhancing RAM:DISK ratio?

**Comment by mykaul** [comment]
> @mykaul , why was this issue bumped to P1? Is it meant to be in Q2 content?

Yes, we should start working on it at high priority.

**Comment by denesb** [comment]
@mykaul @bhalevy which team will work on this?

**Comment by denesb** [comment]
Replaced by SCYLLADB-118.
